tests/test_client_functional.py

```diff
@@ -2333,25 +2335,74 @@ async def test_creds_in_auth_and_url() -> None:
         await session.close()
 
 
-async def test_drop_auth_on_redirect_to_other_host(aiohttp_server: Any) -> None:
+@pytest.mark.parametrize(
+    ["url_a", "url_b", "drop_header"],
+    [
+        [
+            "http://host1.com/path1",
+            "http://host2.com/path2",
+            True,
+        ],  # entirely different hosts
+        ["http://host1.com/path1", "https://host1.com/path1", False],  # http -> https
+        ["https://host1.com/path1", "http://host1.com/path2", True],  # https -> http
+    ],
+)
+async def test_drop_auth_on_redirect_to_other_host(
+    aiohttp_server: Any,
+    tls_certificate_authority: Any,
+    url_a: str,
+    url_b: str,
+    drop_header: bool,
+) -> None:
+    yarl_a = URL(url_a)
+    yarl_b = URL(url_b)
+
     async def srv1(request):
-        assert request.host == "host1.com"
+        assert request.host == yarl_a.host
         assert request.headers["Authorization"] == "Basic dXNlcjpwYXNz"
-        raise web.HTTPFound("http://host2.com/path2")
+        raise web.HTTPFound(url_b)
 
     async def srv2(request):
-        assert request.host == "host2.com"
-        assert "Authorization" not in request.headers
+        assert request.host == yarl_b.host
+        if drop_header:
+            assert "Authorization" not in request.headers, "Header wasn't dropped"
+        else:
+            assert "Authorization" in request.headers, "Header was dropped"
         return web.Response()
 
-    app = web.Application()
-    app.router.add_route("GET", "/path1", srv1)
-    app.router.add_route("GET", "/path2", srv2)
+    async def create_server(url: URL, srv: Any):
+        app = web.Application()
+        app.router.add_route("GET", url.path, srv)
 
-    server = await aiohttp_server(app)
+        if url.scheme == "https":
+            cert = tls_certificate_authority.issue_cert(
+                url.host, "localhost", "127.0.0.1"
+            )
+            ssl_ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            cert.configure_cert(ssl_ctx)
+            kwargs = dict(ssl=ssl_ctx)
+        else:
+            kwargs = {}
+
+        return await aiohttp_server(app, **kwargs)
+
+    server_a = await create_server(yarl_a, srv1)
+    server_b = await create_server(yarl_b, srv2)
+
+    assert (
+        yarl_a.host != yarl_b.host or server_a.scheme != server_b.scheme
+    ), "Invalid test case, host or scheme must differ"
+
+    schemes = dict(http=80, https=443)
```

Comment:
plz use a dict literal here, not a constructor.

Reply:
Also, it's s/schemes/schemas/ in this case.

Reply:
I think it's better to reflect what this variable is in its name: `protocol_port_map`

Reply:
Done.

By the way, why dict literal is preferable?

Reply:
It's slightly faster but also it's because of the consistent style considerations. Do you do `list()` everywhere instead of `[]`? How about `tuple()` vs `()`? The only case I'd use `dict` is when I need to pass callable as an argument.